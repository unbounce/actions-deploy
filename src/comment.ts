import Parsimmon from "parsimmon";

export const details = (summary: string, body: string) => {
  return `<details>\n<summary>${summary}</summary>\n\n${body}\n\n</details>`;
};

export const mention = (body: string) => {
  return `@${process.env.GITHUB_ACTOR} ${body}`;
};

export const codeBlock = (body: string) => {
  const ticks = "```";
  return `${ticks}\n${body}\n${ticks}`;
};

export const code = (body: string) => {
  const tick = "`";
  return `${tick}${body}${tick}`;
};

export const runLink = (text: string) => {
  const url = `https://github.com/${process.env.GITHUB_REPOSITORY}/actions/runs/${process.env.GITHUB_RUN_ID}?check_suite_focus=true`;
  return `[${text}](${url})`;
};

const createLogToDetailsParser = () => {
  const startRe = /::group::(\w+)?/;
  const startGroup = Parsimmon.regex(startRe, 1);
  const endRe = /::endgroup::\s*/;
  const endGroup = Parsimmon.regex(endRe);
  const textThenEndgroup = Parsimmon.regex(/[\s\S]*?(?=::endgroup::\s*)/).trim(
    Parsimmon.optWhitespace
  );
  const textThenEnd = Parsimmon.regex(/[\s\S]*?(?!::endgroup::\s*)$/).trim(
    Parsimmon.optWhitespace
  );
  const groupParser = Parsimmon.seqMap(
    Parsimmon.alt(
      Parsimmon.seq(startGroup, textThenEndgroup, endGroup),
      Parsimmon.seq(startGroup, textThenEnd, Parsimmon.end)
    ),
    ([start, middle, _end]) => {
      const name = start ? start : "Details";
      return details(name, codeBlock(middle));
    }
  ).trim(Parsimmon.optWhitespace);
  return groupParser.many();
};
const logToDetailsParser = createLogToDetailsParser();

// Parse log output into <details> blocks based on GitHub Actions group
// annotations.
//
// For example:
//
// ::group::Name
// One
// Two
// ::endgroup::
//
// will be converted to:
//
// <details>
// <summary>Name</summary>
// ```
// One
// Two
// ```
// </details>
//
// Supports multiple groups in succession and the omission of a final
// ::endgroup:: (in the case that the running script was not able to emit it do
// to an earlier failure).
//
// If any text is found outside of a group, or if no groups are used, the log
// output will be returned as a single code block.
//
export const logToDetails = (log: string) => {
  const parsed = logToDetailsParser.parse(log);
  if (parsed.status) {
    return parsed.value.join("\n");
  } else {
    return codeBlock(log);
  }
};
