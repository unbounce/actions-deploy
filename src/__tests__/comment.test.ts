import { logToDetails } from "../comment";

describe("logToDetails", () => {
  test("with no groups", () => {
    const log = "one\ntwo\nthree";
    expect(logToDetails(log)).toEqual(`\`\`\`\n${log}\n\`\`\``);
  });

  test("with log outside group", () => {
    const log = "before\n::group::One\none\n::endgroup::";
    expect(logToDetails(log)).toEqual(`\`\`\`\n${log}\n\`\`\``);
  });

  test("one group", () => {
    const log = "::group::One\none\n::endgroup::";
    const details = logToDetails(log);
    expect(details.includes("::")).toBeFalsy();
    expect(details).toMatchSnapshot();
  });

  test("no name", () => {
    const log = "::group::\none\n::endgroup::";
    const details = logToDetails(log);
    expect(details.includes("::")).toBeFalsy();
    expect(details).toMatchSnapshot();
  });

  test("many groups", () => {
    const log =
      "::group::One\none\n::endgroup::\n::group::Two\ntwo\n::endgroup::";
    const details = logToDetails(log);
    expect(details.includes("::")).toBeFalsy();
    expect(details).toMatchSnapshot();
  });

  test("many groups with whitespace", () => {
    const log =
      "::group::One\none\n::endgroup::\n \n::group::Two\ntwo\n::endgroup::";
    const details = logToDetails(log);
    expect(details.includes("::")).toBeFalsy();
    expect(details).toMatchSnapshot();
  });

  test("with special characters", () => {
    const log =
      "::group::One\none\n1\n-->>--=\n::endgroup::\n::group::Two\ntwo\n::endgroup::";
    const details = logToDetails(log);
    expect(details.includes("::")).toBeFalsy();
    expect(details).toMatchSnapshot();
  });

  test("missing endgroup", () => {
    const log = "::group::One\none\n::endgroup::\n::group::Two\ntwo\n";
    const details = logToDetails(log);
    expect(details.includes("::")).toBeFalsy();
    expect(details).toMatchSnapshot();
  });
});
