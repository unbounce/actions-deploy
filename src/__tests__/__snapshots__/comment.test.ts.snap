// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logToDetails many groups 1`] = `
"<details>
<summary>One</summary>

\`\`\`
one

\`\`\`

</details>
<details>
<summary>Two</summary>

\`\`\`
two

\`\`\`

</details>"
`;

exports[`logToDetails many groups with whitespace 1`] = `
"<details>
<summary>One</summary>

\`\`\`
one

\`\`\`

</details>
<details>
<summary>Two</summary>

\`\`\`
two

\`\`\`

</details>"
`;

exports[`logToDetails missing endgroup 1`] = `
"<details>
<summary>One</summary>

\`\`\`
one

\`\`\`

</details>
<details>
<summary>Two</summary>

\`\`\`
two

\`\`\`

</details>"
`;

exports[`logToDetails no name 1`] = `
"<details>
<summary>Details</summary>

\`\`\`
one

\`\`\`

</details>"
`;

exports[`logToDetails one group 1`] = `
"<details>
<summary>One</summary>

\`\`\`
one

\`\`\`

</details>"
`;

exports[`logToDetails with special characters 1`] = `
"<details>
<summary>One</summary>

\`\`\`
one
1
-->>--=

\`\`\`

</details>
<details>
<summary>Two</summary>

\`\`\`
two

\`\`\`

</details>"
`;
